pyresample/bilinear/__init__.py

```diff
@@ -210,366 +200,706 @@ def get_bil_info(source_geo_def, target_area_def, radius=50e3, neighbours=32,
-    # Read output coordinates
-    out_lons, out_lats = target_area_def.get_lonlats()
 
-    # Replace masked arrays with np.nan'd ndarrays
-    out_lons = _convert_masks_to_nans(out_lons)
-    out_lats = _convert_masks_to_nans(out_lats)
+def _resample(corner_points, fractional_distances):
+    p_1, p_2, p_3, p_4 = corner_points
+    s__, t__ = fractional_distances
+    return (p_1 * (1 - s__) * (1 - t__) +
+            p_2 * s__ * (1 - t__) +
+            p_3 * (1 - s__) * t__ +
+            p_4 * s__ * t__)
 
-    # Mask invalid coordinates
-    out_lons, out_lats = _mask_coordinates(out_lons, out_lats)
 
-    # Convert coordinates to output projection x/y space
-    out_x, out_y = proj(out_lons, out_lats)
+def query_no_distance(target_lons, target_lats,
+                      valid_output_index, kdtree, neighbours, epsilon, radius):
+    """Query the kdtree. No distances are returned."""
+    voir = np.ravel(valid_output_index)
+    target_lons_valid = np.ravel(target_lons)[voir]
+    target_lats_valid = np.ravel(target_lats)[voir]
 
-    return out_x, out_y
+    _, index_array = kdtree.query(
+        lonlat2xyz(target_lons_valid, target_lats_valid),
+        k=neighbours,
+        eps=epsilon,
+        distance_upper_bound=radius)
 
+    return index_array
 
-def _get_input_xy(source_geo_def, proj, input_idxs, idx_ref):
-    """Get x/y coordinates for the input area and reduce the data."""
-    in_lons, in_lats = source_geo_def.get_lonlats()
 
-    # Select valid locations
-    in_lons = in_lons.ravel()[input_idxs]
-    in_lats = in_lats.ravel()[input_idxs]
+class NumpyResamplerBilinear(BilinearBase):
+    """Bilinear interpolation using Numpy."""
 
-    # Mask invalid values
-    in_lons, in_lats = _mask_coordinates(in_lons, in_lats)
+    def get_sample_from_bil_info(self, data, fill_value=None, output_shape=None):
+        """Resample using pre-computed resampling LUTs."""
+        del output_shape
 
-    # Expand input coordinates for each output location
-    in_lons = in_lons[idx_ref]
-    in_lats = in_lats[idx_ref]
+        res = _resample(
+            self._slice_data(data, fill_value),
+            (self.bilinear_s, self.bilinear_t)
+        )
 
-    # Replace masked arrays with np.nan'd ndarrays
-    in_lons = _convert_masks_to_nans(in_lons)
-    in_lats = _convert_masks_to_nans(in_lats)
+        return self._finalize_output_data(data, res, fill_value)
 
-    # Convert coordinates to output projection x/y space
-    in_x, in_y = proj(in_lons, in_lats)
+    def _slice_data(self, data, fill_value):
+        data = _check_data_shape(data, self._valid_input_index)
+        res = get_slicer(data)(data, self.slices_x, self.slices_y, self.mask_slices, fill_value)
 
-    return in_x, in_y
+        return res
 
+    def _finalize_output_data(self, data, res, fill_value):
+        return _apply_fill_value_or_mask_data(
+            self._reshape_to_target_area(res, data.ndim),
+            fill_value
+        )
```

Comment:
Taking the inline call of `_reshape_to_target_area` out would make this clearer in my opinion. So `_finalize_output_data` would be a two step method (reshape, fill/mask).

Reply:
Yep, adjusted.